pkg/importer/dataStream.go

```diff
@@ -915,51 +845,38 @@ func (d *DataStream) convertQcow2ToRaw(src, dest string) error {
 }
 
 // Copy endpoint to dest based on passed-in reader.
-func (d *DataStream) copy(dest string) (err error) {
+// if the error is not nil and true is return, we require scratch space to complete the copy.
+func (d *DataStream) copy(dest string, hasScratch bool, scratchDataDir string) (bool, error) {
 	if d.isHTTPQcow2() {
-		err = d.convertQcow2ToRawStream(dest)
+		err := d.convertQcow2ToRawStream(dest)
 		if err != nil {
-			return err
+			return false, err
 		}
 	} else {
-		dest = filepath.Clean(dest)
-		glog.V(2).Infof("copying image file to %q", dest)
-		tmpDest := dest
-		if !d.isQemuStored() {
-			//it is either not qemu or not stored
-			if d.qemu {
-				// copy to tmp; qemu conversion will write to passed-in destination
-				tmpDest = randTmpName(dest)
-				glog.V(3).Infof("Copy: temp file for qcow2 conversion: %q", tmpDest)
-				defer func(f string) {
-					os.Remove(f)
-				}(tmpDest)
-			}
-			if d.isMoveable() {
-				glog.V(3).Infof("Moving already stored raw from %s to %s", d.tmpDataPath, tmpDest)
-				err = MoveFile(d.tmpDataPath, tmpDest)
-				if err != nil {
-					glog.V(3).Infof("Failing back to copying stored raw from %s to %s", d.tmpDataPath, tmpDest)
-					//Failback to copy since move failed
-					err = StreamDataToFile(d.topReader(), tmpDest)
-				}
-			} else {
-				glog.V(3).Infof("Streaming data of non Movble file to %s", tmpDest)
-				err = StreamDataToFile(d.topReader(), tmpDest)
-			}
-			if err != nil {
-				return errors.WithMessage(err, fmt.Sprintf("unable to stream data to file %q", dest))
-			}
-		} else {
-			//It is a stored qemu file - taking a location where it is stored
-			glog.V(3).Infof("Streaming file to %s", tmpDest)
-			tmpDest = d.tmpDataPath
+		if !hasScratch {
+			//Need scratch space but none provided.
+			return true, errors.New("Scratch space required, and none found")
+		}
+		// Replace /data/target name with scratch path/target name
+		tmpDest := filepath.Join(scratchDataDir, filepath.Base(dest))
+
+		err := StreamDataToFile(d.topReader(), tmpDest)
+		if err != nil {
+			return false, err
```

Comment:
In this case CleanDir is not called - this will leave files on /data. 
Why the existing solution in registry code has to be removed - it handles all theses cases. Please see above